pydoc-markdown/src/pydoc_markdown/contrib/renderers/docusaurus.py

```diff
@@ -0,0 +1,129 @@
+# -*- coding: utf8 -*-
+from nr.databind.core import Field, Struct
+from nr.interface import implements, override
+from pathlib import Path
+from pydoc_markdown.interfaces import Renderer
+from pydoc_markdown.contrib.renderers.markdown import MarkdownRenderer
+from typing import Any, Dict, List, Text
+import docspec
+import json
+import os.path
+
+
+@implements(Renderer)
+class DocusaurusRenderer(Struct):
+  """
+  Produces Markdown files to be used with Docusaurus websites.
+
+  ### Options
+  """
+
+  #: The #MarkdownRenderer configuration.
+  markdown = Field(
+    MarkdownRenderer,
+    default=MarkdownRenderer(
+      # disabled because Docusaurus supports this automatically
+      insert_header_anchors=False,
+      # escape html in docstring, otherwise it could lead to invalid html
+      escape_html_in_docstring=True,
+      # do not generate any page for empty modules
+      skip_empty_modules=True,
+      # conforms to Docusaurus header format
+      render_module_header_template=(
+        '---\n'
+        'sidebar_label: {module_name}\n'
+        'title: {module_name}\n'
+        '---\n\n'
+      )
+    ),
+  )
+
+  #: The path where the docusaurus docs content is. Defaults "docs" folder.
+  docs_base_path = Field(str, default='docs')
+
+  #: The output path inside the docs_base_path folder, used to ouput the
+  #: module reference.
+  relative_output_path = Field(str, default='reference')
+
+  #: The sidebar path inside the docs_base_path folder, used to ouput the
+  #: sidebar for the module reference.
+  relative_sidebar_path = Field(str, default='sidebar.json')
+
+  #: The top-level label in the sidebar. Default to 'Reference'.
+  sidebar_top_level_label = Field(str, default='Reference')
+
+  @override
+  def render(self, modules: List[docspec.Module]) -> None:
+    modules_and_paths = []
+    module_tree = {"children": {}, "edges": []}
+    output_path = Path(self.docs_base_path) / self.relative_output_path
+    for module in modules:
+      filepath = output_path
+
+      module_parts = module.name.split(".")
+      relative_module_tree = module_tree
+      intermediary_module = []
+
+      for module_part in module_parts[:-1]:
+        # update the module tree
+        intermediary_module.append(module_part)
+        intermediary_module_name = ".".join(intermediary_module)
+        relative_module_tree["children"].setdefault(intermediary_module_name, {"children": {}, "edges": []})
+        relative_module_tree = relative_module_tree["children"][intermediary_module_name]
+
+        # descend to the file
+        filepath = filepath / module_part
+
+      # create intermediary missing directories and get the full path
+      filepath.mkdir(parents=True, exist_ok=True)
+      filepath = filepath / f"{module_parts[-1]}.md"
+
+      with filepath.open('w') as fp:
+        # FIXME: not the cleanest way of doing this...
```

Comment:
True, I was considering adding a `render_to_stream()` function in the Markdown Renderer. The Hugo renderer does the same thing at the moment.

Reply:
c04cbd4 ✅